app/models/location/suggestions.rb

```diff
@@ -0,0 +1,281 @@
+# frozen_string_literal: true
+
+# Return array of suggested Locations for a given user input (name or lat/lon)
+class Location < AbstractModel
+  # SQL string to calculate db grid size in lat/lon squares
+  # This is the db counterpart of pseudoarea, differing by rounding error
+  GRID_SQUARES = "IF(east >= west, east - west, 360 + east - west) " \
+                 "* (north - south)"
+
+  # max suggested size in lat/lon "squares"
+  # suggestions should not include a Location with a larger pseudoarea
+  MAX_SUGGESTED_SIZE = 6800
+
+  # class methods
+  class << self
+    # Return some locations that are close to the given name and/or geolocation
+    # data from google and/or lat/long provided by user.
+    def suggestions(str, geolocation)
+      return suggestions_for_country(str, geolocation) if dubious_country?(str)
+      return suggestions_for_state(str, geolocation) if dubious_state?(str)
+
+      suggestions_for_the_rest(str)
+    end
+
+    # Turn geolocation structure into an MO location name (might not exist).
+    def geolocation_to_name(geolocation)
+      geo_country = geolocation[:country]
+      geo_state = geolocation[:state]
+      geo_county = geolocation[:county].to_s.sub(/County$/, "Co.")
+      geo_city = geolocation[:city]
+      [geo_city, geo_county, geo_state, geo_country].reject(&:blank?).join(", ")
+    end
+
+    # Locations with enough verbal levels of precision
+    # containing the given lat/lon, excluding Locations that are "too large"
+    # sorted by size, ascending
+    def suggestions_for_latlong(lat, long)
+      all = Location.where("(south <= ? AND north >= ?) AND " \
+                           "IF(east >= west, west <= ? AND east >= ?, " \
+                           "west <= ? OR east >= ?)",
+                           lat, lat, long, long, long, long).
+            # exclude locations that are too large
+            where("#{GRID_SQUARES} <= #{MAX_SUGGESTED_SIZE}").
+            to_a
+      # Return only most "precise" Locations.
+      [3, 2, 1, 0].each do |n|
+        locs = all.select { |loc| loc.name.split(",").length > n }
+        return locs.sort_by(&:pseudoarea) if locs.any?
```

Comment:
#### Is there a reasonable way to get the db to sort the results by size ascending?
I'd rather have that done by the db than by Rails. But I couldn't figure out the right syntax. Maybe it's so complex that we should just leave it to Rails. There won't be a noticeable performance difference if there aren't many "precise" hits.

I found it helpful to have the suggestions sorted by size. Compare unsorted (good suggestions highlighted):
![Screen Shot 2020-12-27 at 16 01 06](https://user-images.githubusercontent.com/45460/103182142-f4bb3500-485d-11eb-8415-627f41dbfd5c.jpg)
to sorted:
![Screen Shot 2020-12-27 at 16 00 11](https://user-images.githubusercontent.com/45460/103182120-bc1b5b80-485d-11eb-8380-df7e0854eace.jpg)